### PR TITLE
Add filter is minimised to a single button in active filters

### LINKED ISSUE #592: [Feature Request]: UI to add filter should be minimalized to a single button
### Feature description

Instead of a separate panel make a small buttons which dropsdown a pickerInput (one per dataset including MAE object)

Skärmavbild 2024-05-22 kl  20 23 59

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### Contribution Guidelines

- [X] I agree to follow this project's Contribution Guidelines.

### Security Policy

- [X] I agree to follow this project's Security Policy.

### PR DESCRIPTION
**Pull request by vedhav**
Closes #592

Please checkout to these branches before testing:
1. **teal** - `669_insertUI@main`
2. **teal.slice** - `remove_add_filters_ui@669_insertUI@main`

## Example  shiny app to test:
```r
pkgload::load_all("teal.slice")
library(shinyjs)


# create a FilteredData object
data(miniACC, package = "MultiAssayExperiment")
datasets <- init_filtered_data(list(iris = iris, mtcars = mtcars, MAE = miniACC))

# setting initial state
set_filter_state(
  datasets = datasets,
  filter = teal_slices(
    teal_slice(dataname = "iris", varname = "Species", selected = "virginica", keep_na = FALSE),
    teal_slice(dataname = "mtcars", id = "4 cyl", title = "4 Cylinders", expr = "cyl == 4"),
    teal_slice(dataname = "mtcars", varname = "mpg", selected = c(20.0, 25.0), keep_na = FALSE, keep_inf = FALSE),
    teal_slice(
      dataname = "MAE", varname = "years_to_birth", selected = c(30, 50), keep_na = TRUE
    ),
    teal_slice(
      dataname = "MAE", varname = "vital_status", selected = "1", keep_na = FALSE
    ),
    teal_slice(
      dataname = "MAE", varname = "gender", selected = "female", keep_na = TRUE
    ),
    teal_slice(
      dataname = "MAE", varname = "ARRAY_TYPE", selected = "", keep_na = TRUE
    )
  )
)

ui <- fluidPage(
  shinyjs::useShinyjs(),
  fluidRow(
    column(
      width = 9,
      tabsetPanel(
        tabPanel(title = "iris", dataTableOutput("iris_table")),
        tabPanel(title = "mtcars", dataTableOutput("mtcars_table"))
      )
    ),
    # ui for the filter panel
    column(width = 3, datasets$ui_filter_panel("filter_panel"))
  )
)

server <- function(input, output, session) {
  # this is the shiny server function for the filter panel and the datasets
  # object can now be used inside the application
  datasets$srv_filter_panel("filter_panel")

  # get the filtered datasets and put them inside reactives for analysis
  iris_filtered_data <- reactive(datasets$get_data(dataname = "iris", filtered = TRUE))
  mtcars_filtered_data <- reactive(datasets$get_data(dataname = "mtcars", filtered = TRUE))

  output$iris_table <- renderDataTable(iris_filtered_data())
  output$mtcars_table <- renderDataTable(mtcars_filtered_data())
}

shinyApp(ui, server)
```

## Example teal app to test
```r
pkgload::load_all("teal.slice")
pkgload::load_all("teal")
library(teal.modules.general)


# create a FilteredData object
data(miniACC, package = "MultiAssayExperiment")

data <- teal_data(
  iris = iris,
  mtcars = mtcars,
  MAE = miniACC
)

app <- init(
  data = data,
  modules = tm_variable_browser()
)

shinyApp(app$ui, app$server)
```

### CODE REVIEW

**Comment by vedhav** [comment]
I see that the teal app fails. I'm looking into this now.

**Comment by vedhav** [comment]
Nevermind, It works fine. I was in the wrong `teal` branch 🤦🏽

File: R/FilteredDataset.R [hunk 1]
```diff
@@ -231,6 +231,20 @@ FilteredDataset <- R6::R6Class( # nolint
...
+                  class = "remove pull-right"
+                ),
+                shinyjs::hidden(
+                  actionLink(
+                    ns("hide_new_filter_ui"),
+                    label = "",
+                    icon = icon("minus", lib = "font-awesome"),
+                    class = "remove pull-right"
+                  )
+                ),
```
Comment (gogonzo) [inline R/FilteredDataset.R:247]: I think we can have a single button which toggles `add` panel with consistent naming.

button: `show_hide_add_panel`
panel: `add_panel` or just `add`

File: R/FilteredDataset.R [hunk 2]
```diff
@@ -320,6 +343,21 @@ FilteredDataset <- R6::R6Class( # nolint
...
+            shinyjs::toggle("hide_new_filter_ui")
+          })
+
+          observeEvent(input$hide_new_filter_ui, {
+            shinyjs::hide("filter_add_ui")
+            shinyjs::toggle("show_new_filter_ui")
+            shinyjs::toggle("hide_new_filter_ui")
+          })
+
+          self$srv_add(private$dataname)
```
Comment (gogonzo) [inline R/FilteredDataset.R:358]: We need to keep the feature of not showing add panel when in `teal_slices` is specified with `allow_add = FALSE`.

```suggestion
          if (<teal_slices allow_add> == TRUE>) {
          self$ui_add(ns("add"), active_datanames)
        }
```

**Review by gogonzo — CHANGES_REQUESTED** [review]
`allow_add` is ignored.

File: R/FilteredDataset.R (hunk 2 above)
Comment (vedhav) [reply R/FilteredDataset.R:358]: I see that the `allow_add` completely hid the "Add filters" panel before. However, now it is possible to selectively hide the `+` button allowing the user to disable addition for selective datasets.
So, what if `allow_add` took dataset names as input. When not provided all datasets can be added, but, when provided, only those datasets can be added?

Comment (gogonzo) [reply R/FilteredDataset.R:358]: > So, what if allow_add took dataset names as input

@vedhav let's not introduce breaking changes. `allow_add` has specific use cases and it came from end users.

File: R/FilteredData.R [hunk 3]
```diff
@@ -554,7 +548,7 @@ FilteredData <- R6::R6Class( # nolint
     #' @param active_datanames (`reactive`)
     #'   defining subset of `self$datanames()` to be displayed.
     #' @return `shiny.tag`
-    ui_active = function(id, active_datanames = self$datanames()) {
+    ui_active = function(id, active_datanames = self$datanames(), allow_add = TRUE) {
```
Comment (gogonzo) [inline R/FilteredData.R:551]: `allow_add` is not an option for developer - it should be provided through `teal_slices(allow_add = FALSE)`. In `teal` we shouldn't use `allow_add` directly, it should be done automatically here in `teal.slice`

```suggestion
    ui_active = function(id, active_datanames = self$datanames()) {
```

File: R/FilteredData.R [hunk 4]
```diff
@@ -582,10 +576,10 @@ FilteredData <- R6::R6Class( # nolint
           id = ns("filter_active_vars_contents"),
           tagList(
             lapply(
-              active_datanames,
+              isolate(active_datanames()),
               function(dataname) {
                 fdataset <- private$get_filtered_dataset(dataname)
-                fdataset$ui_active(id = ns(dataname))
+                fdataset$ui_active(id = ns(dataname), allow_add)
```
Comment (gogonzo) [inline R/FilteredData.R:582]: ... which means 
```suggestion
                fdataset$ui_active(id = ns(dataname), allow_add = private$allow_add)
```

File: R/FilteredDataset.R [hunk 5]
```diff
@@ -244,6 +256,17 @@ FilteredDataset <- R6::R6Class( # nolint
                   class = "remove pull-right"
                 )
               )
+            ),
+            column(
+              width = 12,
+              if (allow_add) {
+                shinyjs::hidden(
+                  tags$div(
+                    id = ns("filter_add_ui"),
```
Comment (gogonzo) [inline R/FilteredDataset.R:265]:
```suggestion
                    id = ns("add_panel"),
```

File: R/FilteredDataset.R [hunk 6]
```diff
@@ -231,6 +234,15 @@ FilteredDataset <- R6::R6Class( # nolint
             column(
               width = 4,
               tagList(
+                if (allow_add) {
+                  actionLink(
+                    ns("show_hide_add_panel"),
```
Comment (gogonzo) [inline R/FilteredDataset.R:239]:
```suggestion
                    ns("toggle_add_panel"),
```

**Review by gogonzo — COMMENTED** [review]
- I think we can simplify even further (see all the comments below)
- If we remove `add_ui/srv` it should be in the NEWS as it is a public method of the exported class.

**Pull request merged**